### DISCUSSION TITLE
Running certain workloads in a "pseudo-airgapped" environment - best way to approach dependency updates?

### DISCUSSION THREAD

**Discussion by jlehman9** [renovatebot/renovate#25241] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Latest 

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Basically I have an environment where we host GitLab and Renovate bot. For all intents and purposes this environment is airgapped, but I can have a non-airgrpped GitLab runner connect to this GitLab instance. Theoretically we could have Renovate bot run against our repos as they stand - but due to certain requirements our repos aren't supposed to leave the network. My understanding is that when Renovate runs it does a full clone of the repo which means the default method is a non-starter. 

Ideally we would have something like Artifactory setup where we can airgap the dependencies or have a pull-trhough system but that isn't available to me. 

Some ideas I've toyed around with:
- Have each repo contain a branch that only has the dependency manifest(s) and the renovate config and run renovate against that branch. Unforunately that doesn't seem possible according to [this discussion](https://github.com/renovatebot/renovate/discussions/16108).
- Have a separate repo only containing the dep. manifests of various other repos and either:
    - Have one instance of Renovate run against the "combined dep" repo on the outside GitLab runner, then have another instance of Renovate run against the repos inside the network using the cache of the dependency repo to know what needs updating and creating MRs for each repo. Renovate doesn't have this functionality AFAIK and I'd rather not make & maintain a custom build for it.
    - Have GitLab automation run on the "combined deps" repo to package all of the dependencies into container images and builds apps with the images. This method _works_ but the Renovate bot MRs would only appear on this shared dep repo and not the individual repos. We would then need a way to update the dependencies in each project's repo after they been updated in the shared repo.
    
Is there something obvious I'm missing that would solve this issue for me? The biggest "want" is having a way for Renovate bot to make the MRs on the project repos without pulling the actual project code outside of the airgap. Hopefully that makes sense

### What you tried so far.

I had a mostly working instance of the last bullet option from above, but the lack of ability to have Renovate make MRs on the original projects vs. having them all made on the "shared deps" repo makes it really difficult to use in practice. Not only in terms of initial setup/configuration for each repo, but also maintaining each repo with each update.

**Comment by viceice** [comment]
you can use dynamic install mode of renovate

https://github.com/containerbase/base/blob/main/docs/custom-registries.md#custom-registries

sadly I've not yet documented all urls, but you can simply run renovate and check the logs for failed tool downloads.

**Reply by jlehman9** [reply]
I'm sorry, can you explain how custom registries are relevant? Maybe I wasn't clear in my explanation of my problem or I'm not understanding your solution.

**Reply by rarkins** [reply]
If you use the default binarySource=install (check docs) then you either need internet access or have mapped each binary source to a local proxy.

Otherwise use binarySource=global
